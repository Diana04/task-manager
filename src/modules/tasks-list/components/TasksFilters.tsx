import { useCallback } from 'react';
import { useShallow } from 'zustand/shallow';
import { Checkbox, Flex, Typography } from 'antd';
import styled from 'styled-components';

import { useTasksStore } from '../store';

const StyledTypographyTitle = styled(Typography.Title)`
  margin: 0 !important;
`;

export const TasksFilters = () => {
    const {taskTags, activeFilters, setActiveFilters} = useTasksStore(useShallow(state => ({
        taskTags: state.tags,
        activeFilters: state.activeFilters,
        setActiveFilters: state.setActiveFilters,
    })));
    
    const handleChange = useCallback((tagIds: string[]) => {
        setActiveFilters({ tagIds })
    }, [setActiveFilters]);

    return (
        <Flex gap="small" vertical>
            <StyledTypographyTitle level={5}>Filters</StyledTypographyTitle>
            <Flex gap="middle">
                <Checkbox.Group onChange={handleChange} value={activeFilters.tagIds}>
                    {taskTags && taskTags.map((tag) => (
                        <Checkbox value={tag.id} key={tag.id}>
                            {tag.title}
                        </Checkbox>
                    ))}
                </Checkbox.Group>
            </Flex>
        </Flex>
    );
}