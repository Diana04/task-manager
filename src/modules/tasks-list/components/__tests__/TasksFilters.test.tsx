import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { TasksFilters } from '../TasksFilters';

const setActiveFiltersMock = jest.fn();

jest.mock('antd', () => ({
    ...jest.requireActual('antd'),
    Flex: ({ children }: { children: any }) => <>{children}</>,
    Typography: {
        Title: ({ children }: { children: any }) => <>{children}</>,
    },
}));

jest.mock('../../store', () => ({
    useTasksStore: (selector: any) =>
        selector({
            tags: [
                {
                    id: 'tag-1',
                    title: 'tag-1',
                    color: 'color-1',
                },
                {
                    id: 'tag-2',
                    title: 'tag-2',
                    color: 'color-2',
                },
            ],
            activeFilters: {
                tagIds: ['tag-2'],
            },
            setActiveFilters: setActiveFiltersMock,
        }),
}));

describe('TasksFilters', () => {
    describe('snapshot check', () => {
        it('should render correctly according to snapshot', () => {
            const { container } = render(<TasksFilters />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('render check', () => {
        it('should render 2 checkboxes', () => {
            render(<TasksFilters />);
            const checkbox = screen.getAllByRole('checkbox');
            expect(checkbox).toHaveLength(2);
        });

        it('should render one unchecked checkbox and one checked checkbox', () => {
            render(<TasksFilters />);
            const checkbox1 = screen.getByDisplayValue('tag-1');
            const checkbox2 = screen.getByDisplayValue('tag-2');

            expect(checkbox1).toBeInTheDocument();
            expect(checkbox1).not.toBeChecked();

            expect(checkbox2).toBeInTheDocument();
            expect(checkbox2).toBeChecked();
        });
    });

    describe('behaviour check', () => {
        it('should call setActiveFilters on checkbox click with new tagsIds list', async () => {
            render(<TasksFilters />);
            const checkbox = screen.getByDisplayValue('tag-1');
            await userEvent.click(checkbox);

            expect(setActiveFiltersMock).toHaveBeenCalledWith({
                tagIds: ['tag-2', 'tag-1'],
            });
        });

        it('should call setActiveFilters on checkbox click with empty tagsIds list', async () => {
            render(<TasksFilters />);
            const checkbox = screen.getByDisplayValue('tag-2');
            await userEvent.click(checkbox);

            expect(setActiveFiltersMock).toHaveBeenCalledWith({ tagIds: [] });
        });
    });
});
