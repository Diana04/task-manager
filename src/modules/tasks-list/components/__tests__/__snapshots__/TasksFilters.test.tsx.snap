// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TasksFilters snapshot check should render correctly according to snapshot 1`] = `
<div>
  Filters
  <div
    class="ant-checkbox-group css-dev-only-do-not-override-l9pxc0"
  >
    <label
      class="ant-checkbox-wrapper css-dev-only-do-not-override-l9pxc0"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-l9pxc0"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="tag-1"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span
        class="ant-checkbox-label"
      >
        tag-1
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-wrapper-checked css-dev-only-do-not-override-l9pxc0"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-l9pxc0 ant-checkbox-checked"
      >
        <input
          checked=""
          class="ant-checkbox-input"
          type="checkbox"
          value="tag-2"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span
        class="ant-checkbox-label"
      >
        tag-2
      </span>
    </label>
  </div>
</div>
`;
